### PR TITLE
Erdikose/feature/getting started wizard

### PR DESCRIPTION
**Pull request by erdkse**
This PR...

## Changes

- adds getting started wizard
- adds projects approach

## Fixes

-

## How to test it

-

## Screenshots

-

## Checklist

- [ ] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: src/redux/selectors.ts [hunk 1]
```diff
@@ -57,11 +59,138 @@ export const isInPreviewModeSelector = createSelector(
 
 export const isInClusterModeSelector = createSelector(
   (state: RootState) => state,
-  appState =>
-    Boolean(appState.main.previewResourceId && appState.main.previewResourceId.endsWith(appState.config.kubeconfigPath))
+  ({main, config}) => {
+    const kubeConfigPath = config.projectConfig?.kubeConfig?.path || config.kubeConfig.path;
+    if (kubeConfigPath) {
+      return Boolean(main.previewResourceId && main.previewResourceId.endsWith(kubeConfigPath));
```
Comment (andreiv1992) [inline src/redux/selectors.ts:65]: Shouldn't this be a boolean by default? since we are using the && operator

File: src/redux/selectors.ts [hunk 2]
```diff
@@ -57,11 +59,138 @@ export const isInPreviewModeSelector = createSelector(
...
+export const mergeConfigs = (baseConfig: ProjectConfig, config?: ProjectConfig | null) => {
+  if (!(baseConfig && baseConfig.settings && baseConfig.kubeConfig)) {
+    throw Error('Base config must be set');
+  }
+
+  if (!config) {
+    return baseConfig;
+  }
+
+  if (
```
Comment (andreiv1992) [inline src/redux/selectors.ts:100]: Adding new options here will add new code and new complexity. Do you think it is a good idea to create config where we map each key and a handler or something similar that we should do. All boolean/string... specific types seem the same logic with different keys

File: src/redux/services/projectConfigMonitor.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      readApplicationConfigFileAndUpdateProjectSettings(absolutePath, dispatch);
+    })
+    .on('change', () => {
+      readApplicationConfigFileAndUpdateProjectSettings(absolutePath, dispatch);
+    })
+    .on('unlink', () => {
+      readApplicationConfigFileAndUpdateProjectSettings(absolutePath, dispatch);
+    });
+
+  watcher
```
Comment (andreiv1992) [inline src/redux/services/projectConfigMonitor.ts:44]: Is there any reason why this cannot be moved to the object on line 33?

File: src/redux/selectors.ts (hunk 1 above)
Comment (erdkse) [reply src/redux/selectors.ts:65]: If the second argument is not boolean AND operator doesn't return boolean

File: src/redux/services/projectConfigMonitor.ts (hunk 3 above)
Comment (erdkse) [reply src/redux/services/projectConfigMonitor.ts:44]: Nope, there is no specific reason - will merge them

File: src/redux/selectors.ts (hunk 1 above)
Comment (andreiv1992) [reply src/redux/selectors.ts:65]: true, sorry about that

File: src/App.tsx [hunk 4]
```diff
@@ -41,6 +59,66 @@ const MainContainer = styled.div`
 
 const App = () => {
   const size: Size = useWindowSize();
+  const dispatch = useDispatch();
+  const currentConfig: ProjectConfig = useSelector(currentConfigSelector);
+  const projects: Project[] = useAppSelector(state => state.config.projects);
+  const projectConfig: ProjectConfig | null | undefined = useAppSelector(state => state.config.projectConfig);
+  const fileMap = useAppSelector(state => state.main.fileMap);
```
Comment (devcatalin) [inline src/App.tsx:66]: Since you're only using the root file entry here, this should probably be:
`const rootFileEntry = useAppSelector(state => state.main.fileMap[ROOT_FILE_ENTRY]);`
Because we don't want App.tsx to re-render on every change to `fileMap`, probably only to changes of the root file entry

File: src/App.tsx [hunk 5]
```diff
@@ -41,6 +59,66 @@ const MainContainer = styled.div`
...
+          );
+        }
+      }
+    },
+    DEFAULT_KUBECONFIG_DEBOUNCE,
+    [
+      currentConfig.kubeConfig?.path,
+      dispatch,
+      projectConfig?.kubeConfig?.currentContext,
+      fileMap[ROOT_FILE_ENTRY]?.filePath,
```
Comment (devcatalin) [inline src/App.tsx:119]: replace this with the rootFileEntry, see comment from above

File: src/components/molecules/CheckedResourcesActionsMenu/CheckedResourcesActionsMenu.tsx [hunk 6]
```diff
@@ -66,10 +66,10 @@ const CheckedResourcesActionsMenu: React.FC = () => {
   const dispatch = useAppDispatch();
 
   const checkedResourceIds = useAppSelector(state => state.main.checkedResourceIds);
-  const currentContext = useAppSelector(state => state.config.kubeConfig.currentContext);
   const isInClusterMode = useAppSelector(isInClusterModeSelector);
   const isInPreviewMode = useAppSelector(isInPreviewModeSelector);
   const resourceMap = useAppSelector(state => state.main.resourceMap);
+  const currentConfig = useAppSelector(currentConfigSelector);
```
Comment (devcatalin) [inline src/components/molecules/CheckedResourcesActionsMenu/CheckedResourcesActionsMenu.tsx:72]: similar to rootFileEntry, if you only need the currentContext, it's better to select just that

File: src/components/molecules/ResourceDiff/ResourceDiff.tsx [hunk 7]
```diff
@@ -163,11 +164,20 @@ const ResourceDiff = (props: {
...
-      isClusterPreview: previewType === 'cluster',
-      shouldPerformDiff: true,
-      isInClusterDiff,
-    });
+    applyResource(
+      localResource.id,
+      resourceMap,
+      fileMap,
+      dispatch,
+      String(currenConfig.kubeConfig?.path),
```
Comment (devcatalin) [inline src/components/molecules/ResourceDiff/ResourceDiff.tsx:172]: what is `String` here used for?

File: src/components/organisms/CreateProjectModal/CreateProjectModal.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  const onFinish = (values: {projectName: string; location: string}) => {
+    setFormValues(values);
+    if (uiState.fromTemplate && formStep === 1) {
+      setFormStep(2);
+    }
+  };
+
+  useEffect(() => {
+    console.log('formValues', formValues);
```
Comment (devcatalin) [inline src/components/organisms/CreateProjectModal/CreateProjectModal.tsx:43]: missed a console log here

File: src/components/organisms/CreateProjectModal/CreateProjectModal.tsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  }, [formValues]);
+
+  useEffect(() => {
+    if (uiState.isOpen) {
+      focus();
+    } else {
+      createProjectForm.resetFields();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [uiState]);
```
Comment (devcatalin) [inline src/components/organisms/CreateProjectModal/CreateProjectModal.tsx:57]: do you need this to be called every time the entire `uiState` object is changed?

File: src/components/organisms/CreateProjectModal/CreateProjectModal.tsx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+          <div>Create a Project</div>
+        )
+      }
+      visible={uiState?.isOpen}
+      onCancel={closeModal}
+      footer={[
+        <Button key="cancel" onClick={closeModal}>
+          Discard
+        </Button>,
+        uiState.fromTemplate && formStep === 2 && (
```
[the remark this answers is not in the document]
Comment (devcatalin) [reply src/components/organisms/CreateProjectModal/CreateProjectModal.tsx:88]: or at least a union type, which usually works well with auto-complete..
`type FormStep = 'step-one' | 'step-two';`

File: src/components/organisms/FileTreePane/FileTreePane.tsx [hunk 11]
```diff
@@ -780,7 +780,11 @@ const FileTreePane = () => {
     if (!fileIncludes.some(fileInclude => micromatch.isMatch(path.basename(info.node.key), fileInclude))) {
```
Comment (devcatalin) [inline src/components/organisms/FileTreePane/FileTreePane.tsx:780]: shouldn't this use the currentConfig as well?

File: src/components/organisms/HotKeysHandler/HotKeysHandler.tsx [hunk 12]
```diff
@@ -35,9 +35,10 @@ import featureJson from '@src/feature-flags.json';
 const HotKeysHandler = () => {
   const dispatch = useAppDispatch();
   const mainState = useAppSelector(state => state.main);
-  const configState = useAppSelector(state => state.config);
+  // const configState = useAppSelector(state => state.config);
```
Comment (devcatalin) [inline src/components/organisms/HotKeysHandler/HotKeysHandler.tsx:38]: remove comment

File: src/components/organisms/NavigatorPane/NavigatorPane.styled.tsx [hunk 13]
```diff
@@ -36,10 +36,24 @@ export const TitleBarRightButtons = styled.div`
...
 export const FiltersNumber = styled.div`
   margin-left: 5px;
 `;
+
+interface PlusItem extends ButtonProps {
+  highlighted?: boolean;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export const PlusButton = styled(({highlighted, ...rest}: PlusItem) => <Button {...rest} />)`
```
Comment (devcatalin) [inline src/components/organisms/NavigatorPane/NavigatorPane.styled.tsx:50]: here you don't have to create a function on every line
you can just do something like this:
```
  ${({highlighted}) => `
      border-radius: ${highlighted ? '100%' : 'inherit'} !important;
      color: ${highlighted ? Colors.whitePure : Colors.blue6} !important;
  `};
```
and the same for `&:after`

File: src/components/organisms/PageHeader/PageHeader.tsx [hunk 14]
```diff
@@ -172,17 +186,29 @@ const CLusterStatusText = styled.span<{connected: Boolean}>`
...
-  font-size: 11px;
-  font-weight: 600;
+const CLusterActionText = styled(
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  ({className, children, highlighted}: {className?: string; children?: any; highlighted?: boolean}) => (
+    <span className={className}>{children}</span>
+  )
+)`
+  font-size: 12px;
+  ${({highlighted}) => `font-size: ${highlighted ? '8px' : '12px'} !important`};
```
Comment (devcatalin) [inline src/components/organisms/PageHeader/PageHeader.tsx:200]: same as above, no need to create so many functions

File: src/components/organisms/PageHeader/PageHeader.tsx [hunk 15]
```diff
@@ -218,20 +259,29 @@ const PageHeader = () => {
...
-  const isClusterActionDisabled = !kubeconfigPath || !isKubeconfigPathValid;
 
   const [previewResource, setPreviewResource] = useState<K8sResource>();
   const [previewValuesFile, setPreviewValuesFile] = useState<HelmValuesFile>();
   const [helmChart, setHelmChart] = useState<HelmChart>();
   const dispatch = useAppDispatch();
+  const activeProject = useSelector(activeProjectSelector);
+  const highlightedItems = useAppSelector(state => state.ui.highlightedItems);
+  const currentConfig = useAppSelector(currentConfigSelector);
+  const projectConfig = useAppSelector(state => state.config.projectConfig);
```
Comment (devcatalin) [inline src/components/organisms/PageHeader/PageHeader.tsx:275]: what's the difference between `currentConfig` and `projectConfig` ?

File: src/components/organisms/PageHeader/PageHeader.tsx [hunk 16]
```diff
@@ -322,14 +387,36 @@ const PageHeader = () => {
...
       return <LoadingOutlined />;
     }
-    return <CLusterActionText>LOAD</CLusterActionText>;
-  }, [previewType, previewLoader, isInClusterMode]);
+    return (
+      <CLusterActionText
+        className={highlightedItems.connectToCluster ? 'animated-highlight' : ''}
+        highlighted={highlightedItems.connectToCluster}
+      >
+        LOAD
```
Comment (devcatalin) [inline src/components/organisms/PageHeader/PageHeader.tsx:395]: I would say we should rename this to "LOAD OBJECTS"

File: src/components/organisms/SettingsDrawer/Settings.tsx [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+
+import {ProjectConfig} from '@models/appconfig';
+
+import {useAppDispatch, useAppSelector} from '@redux/hooks';
+import {updateShouldOptionalIgnoreUnsatisfiedRefs} from '@redux/reducers/main';
+import {isInClusterModeSelector} from '@redux/selectors';
+
+import FilePatternList from '@molecules/FilePatternList';
+
+// import {Themes, TextSizes, Languages} from '@models/appconfig';
```
Comment (devcatalin) [inline src/components/organisms/SettingsDrawer/Settings.tsx:30]: remove this

File: src/components/organisms/SettingsDrawer/Settings.tsx [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+      onConfigChange(localConfig);
+    }
+  };
+
+  useEffect(() => {
+    setIsClusterActionDisabled(Boolean(!config?.kubeConfig?.path) || Boolean(!config?.kubeConfig?.isPathValid));
+    setCurrentKubeConfig(config?.kubeConfig?.path);
+  }, [config?.kubeConfig]);
+
+  useEffect(() => {
```
Comment (devcatalin) [inline src/components/organisms/SettingsDrawer/Settings.tsx:119]: could you add a comment to why is this needed?
it was confusing for me, thinking it makes an infinite loop but then I saw `config` was from props..

File: src/components/organisms/SettingsDrawer/Settings.tsx [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+  const onChangeFileIncludes = (patterns: string[]) => {
+    setLocalConfig({...localConfig, fileIncludes: patterns});
+  };
+
+  const onChangeScanExcludes = (patterns: string[]) => {
+    setLocalConfig({...localConfig, scanExcludes: patterns});
+  };
+
+  // const onChangeTheme = (e: RadioChangeEvent) => {
```
Comment (devcatalin) [inline src/components/organisms/SettingsDrawer/Settings.tsx:143]: remove this as well

File: src/components/organisms/SettingsDrawer/Settings.tsx [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+  const [inputRef, focusInput] = useFocus<Input>();
+  const hasUserPerformedClickOnClusterIcon = useAppSelector(state => state.uiCoach.hasUserPerformedClickOnClusterIcon);
+  const wasRehydrated = useAppSelector(state => state.main.wasRehydrated);
+  const [isClusterActionDisabled, setIsClusterActionDisabled] = useState(
+    Boolean(!config?.kubeConfig?.path) || Boolean(!config?.kubeConfig?.isPathValid)
+  );
+  const [currentKubeConfig, setCurrentKubeConfig] = useState(config?.kubeConfig?.path);
+  const isEditingDisabled = uiState.isClusterDiffVisible || isInClusterMode;
+
+  const [localConfig, setLocalConfig] = useState<ProjectConfig | null | undefined>(config);
```
Comment (devcatalin) [inline src/components/organisms/SettingsDrawer/Settings.tsx:106]: wouldn't `projectConfig` be a better name? or `currentProjectConfig` ?
I'm just getting a bit confused about difference between `currentConfig`, `projectConfig` and `localConfig` across the files

File: src/models/appconfig.ts [hunk 21]
```diff
@@ -27,6 +38,35 @@ export enum NewVersionCode {
   Downloaded = 4,
 }
 
+export type KubeConfigContext = {
+  cluster: string;
```
Comment (devcatalin) [inline src/models/appconfig.ts:42]: can you add a comment to this property or maybe a better name?

File: src/redux/services/fileEntry.ts [hunk 22]
```diff
@@ -107,7 +108,8 @@ export function readFiles(
       if (fileIsExcluded(appConfig, fileEntry)) {
         fileEntry.isExcluded = true;
       } else if (getFileStats(filePath)?.isDirectory()) {
-        if (depth === appConfig.folderReadsMaxDepth) {
+        const folderReadsMaxDepth = appConfig.projectConfig?.folderReadsMaxDepth || appConfig.folderReadsMaxDepth;
```
Comment (devcatalin) [inline src/redux/services/fileEntry.ts:111]: shouldn't the `projectConfig` already have a default value for this? (and the other properties...)

File: src/redux/thunks/loadKubeConfig.ts [hunk 23]
```diff
@@ -1,17 +1,15 @@
...
 import {AlertEnum} from '@models/alert';
-import {KubeConfig, KubeConfigContext} from '@models/kubeConfig';
+import {KubeConfig, KubeConfigContext} from '@models/appconfig';
 
 import {setAlert} from '@redux/reducers/alert';
-import {setContexts} from '@redux/reducers/appConfig';
+import {updateProjectKubeConfig} from '@redux/reducers/appConfig';
 
-export const loadContexts = async (configPath: string, dispatch: (action: AnyAction) => void) => {
+export const loadContexts = async (configPath: string, dispatch: any, currentContext?: string) => {
```
Comment (devcatalin) [inline src/redux/thunks/loadKubeConfig.ts:12]: any reason for removing the `(action: AnyAction) => void` type from `dispatch`?

Comment (devcatalin) [reply src/redux/thunks/loadKubeConfig.ts:12]: beware that you cannot dispatch any Thunks from the main process, only plain actions

**Review by devcatalin — CHANGES_REQUESTED** [review]

**Comment by devcatalin** [comment]
I agree with @andreiv1992 on the styled components, we should try to be more consistent on how and where we place them from now on...
I suggest we either have a `styled.tsx` or a `ComponentName.styled.tsx` file next to the component that is being styled.
Then we can do `import * as S from "./styled` and use `<S.SomeContainer />` afterwards.

I also have a question about the use of `String(...)` in multiple places in this PR, is there any reason behind it?

And just overall, you should add some comments explaining the differences between different configs so we don't get confused when we use them..

File: src/components/organisms/SettingsDrawer/Settings.tsx (hunk 18 above)
Comment (erdkse) [reply src/components/organisms/SettingsDrawer/Settings.tsx:119]: Added

File: src/redux/services/fileEntry.ts (hunk 22 above)
Comment (erdkse) [reply src/redux/services/fileEntry.ts:111]: I will explain in a comment at bottom

File: src/components/organisms/SettingsDrawer/Settings.tsx (hunk 20 above)
Comment (erdkse) [reply src/components/organisms/SettingsDrawer/Settings.tsx:106]: I will explain in a comment at bottom

File: src/components/organisms/PageHeader/PageHeader.tsx (hunk 15 above)
Comment (erdkse) [reply src/components/organisms/PageHeader/PageHeader.tsx:275]: I will explain with a comment at the bottom

File: src/App.tsx [hunk 24]
```diff
@@ -41,6 +59,66 @@ const MainContainer = styled.div`
...
   const size: Size = useWindowSize();
+  const dispatch = useDispatch();
+  const currentConfig: ProjectConfig = useSelector(currentConfigSelector);
+  const projects: Project[] = useAppSelector(state => state.config.projects);
+  const projectConfig: ProjectConfig | null | undefined = useAppSelector(state => state.config.projectConfig);
+  const fileMap = useAppSelector(state => state.main.fileMap);
+
+  const onExecutedFrom = useCallback(
+    (_, data) => {
+      const project: Project =
```
[the remark this answers is not in the document]
Comment (erdkse) [reply src/App.tsx:72]: this is for terminal access

File: src/components/molecules/ModalConfirmWithNamespaceSelect/ModalConfirmWithNamespaceSelect.tsx [hunk 25]
```diff
@@ -77,8 +77,8 @@ const ModalConfirmWithNamespaceSelect: React.FC<IProps> = props => {
         return;
       }
       const kc = new k8s.KubeConfig();
-      kc.loadFromFile(kubeconfigPath);
-      kc.setCurrentContext(currentContext || '');
+      kc.loadFromFile(String(currentConfig.kubeConfig?.path));
```
[the remark this answers is not in the document]
Comment (erdkse) [reply src/components/molecules/ModalConfirmWithNamespaceSelect/ModalConfirmWithNamespaceSelect.tsx:80]: this KubeConfig is from a third-party library. To do what you say we need to fork/contribute to that project

**Comment by erdkse** [comment]
Hey folks, thank you for your reviews!

Now`Monokle` has 2 configurations. One is`appConfig` that we all know and the second is a kinda snapshot of`appConfig`. It carries these properties - https://github.com/kubeshop/monokle/blob/erdikose/feature%2Fgetting-started-wizard/src/models/appconfig.ts#L63

What you saw as`currentConfig` is the merged value of these two configs. I combined them in a way that`projectConfig` overrides all values of`appConfig`. After review, I added individual store selectors for most used values like `kubeConfigPath`, `kubeConfigContext` etc. At the places where I couldn't use these selectors, I manually override `projectConfig` over `appConfig`.

If you have more questions please ask me. And I kindly demand your re-reviews 😄

**Review by olensmar — APPROVED** [review]
too big to review for me - ok to merge!

**Review by devcatalin — APPROVED** [review]
Let's merge it and see what happens
